### PR TITLE
pageserver: add `InProgress` tenant map state, use a sync lock for the map

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Follows on from #5299 
- We didn't have a generic way to protect a tenant undergoing changes: `Tenant` had states, but for our arbitrary transitions between secondary/attached, we need a general way to say "reserve this tenant ID, and don't allow any other ops on it, but don't try and report it as being in any particular state".
- The TenantsMap structure was behind an async RwLock, but it was never correct to hold it across await points: that would block any other changes for all tenants.

## Summary of changes

- Add the `TenantSlot::InProgress` value.  This means:
  - Incoming administrative operations on the tenant should retry later
  - Anything trying to read the live state of the tenant (e.g. a page service reader) should retry later or block.
- Store TenantsMap in `std::sync::RwLock`
- Provide an extended `get_active_tenant_with_timeout` for page_service to use, which will wait on InProgress slots as well as non-active tenants.

Closes: https://github.com/neondatabase/neon/issues/5378

### CODE REVIEW

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:53]: Could we use the utils::completion::Barrier here instead?

It has a very naive implementation, but if we could agree to use it (more), then I think it would be good.

Comment (jcsp) [reply pageserver/src/tenant/mgr.rs]: Cut this over to Completion for the RAII convenience of not having to remember to signal the notify :+1:

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:677]: nit: better name for this would be `slot_guard`

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1535]: no naked bools please

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1502]: Why early-exit for `ShuttingDown` here? The `SlotGuard` could have been created earlier while we weren't shutting down.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1514]: Let's follow the same pattern that we had for the deletion queue: have a counter metric for errors that should never happen but we keep going anyway.
(That reminds me: we should configure an alert for that one)

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1520]: The [above](https://github.com/neondatabase/neon/pull/5367#discussion_r1380426149) applies here as well.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:965]: This feels like it should be in `utils` 

Regarding the signature, I think the order `duration, cancel, future` would be better for other use cases where we pass `future = async move { ... }` inline.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1544]: This doesn't match my understanding of the semantics in the doc comment. My understanding from the doc comment is that one should pass `true` when one is ok to operate on a tenant while we're shutting down.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1024]: The better solution here is to compute a deadline at the top of the function, and use `until()` for to compute the timeout arguments to sub-functions. (Really, we should be using deadlines more)

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1043]: Couldn't it become `InProgress` again since we last checked? 

The dirty fix for this would be to turn this function into a loop.

That doesn't ensure fairness, though.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1441]: Would be nice to
1. panic in Drop if we didn't provide a replacement through `upsert`
2. ensure via type system that this can be called only once, i.e., statically ensure all calls return `Some()`

File: pageserver/src/tenant/mgr.rs
Comment (problame) [inline pageserver/src/tenant/mgr.rs:1299]: This seems dangerous. A tenant that is being worked on won't show up in the listing. Can't have that IMO. We should extend the api spec to allow returning a different body than full `TenantState` for tenants that are `InProgress`. Maybe a string that's provided by the acquirer of the slot, indicating why the tenant is `InProgres`. That'd also be a good debugging breadcrumb.

**Review by problame — DISMISSED** [review]

**Comment by problame** [comment]
Overall good job, nice API. Will review another time once above comments are resolved. If you have bigger-picture objects to the review comments, let's talk it over in person.

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1299]: I understand that this is surprising on first glance, but I don't think it is dangerous.

Remember that when we're in this state, we'are always transitioning to or from a state where the tenant would not be visible in the list API:
- while a request to attach a tenant is in progress, we make no promises about that tenant showing up in the external API until we have acked that attach request.
- when we get a request to detach a tenant (or transition to secondary), we do not guarantee the visibility of the tenant throughout that request: there will be some moment in time when that request is still happening (we haven't sent a response), but the tenant is already not present in the listing.

So the API semantic is unchanged:
- while waiting for the response to an attach/detach request, the tenant _might_ be present in the listing
- once the request has been acked, the tenant will be definitively present or absent.

Comment (problame) [reply pageserver/src/tenant/mgr.rs:1299]: Ok, I agree that it's not a semantics change for the existing APIs.

But, it'll be quite odd for `upsert`: when doing upsert in the not-fast-pasth, the tenant briefly disappears.

That being said, I think we should review who is actually using the list endpoint.
I don't think console does.
We should de-publish it from the openapi spec, then I'm not worried anymore.

We should also reduce the published API surface for the per-tenant status endpoint.

Can you do this as a preliminary to this PR? The way to figure out if there's a dependency on the API surface is to make the change in cloud.git first, and if it still compiles, commit the change there, then mirror it in neon.git.

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1041]: This function returns an Arc<Tenant>, so it's always the case that the returned tenant's slot might have gone into an InProgress state by the time the receiver gets it.

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1022]: This is true.  Refactored to use a deadline.  8d6b95b44

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1542]: Fixed in 810404f9

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1533]: 810404f9

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1613]: The exhausive match would be very verbose, as it's all permutations of TenantSlot and the TenantSlotAcquireMode

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1560]: 810404f9

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1500]: The ShuttingDown state never contains InProgress states -- shutdown_all_tenants0 doesn't carry them across.

One could equally copy them across during shutdown and then remove this check, but conceptually it's simpler if we think of the ShuttingDown state as being a readonly view of whatever was present when we started shutting down, rather than being mutable in the same way as the Open state.  Long term, as with other areas, I think we can live without this Open/Initializing/ShuttingDown distinction, and just not let any external APIs hit this object until we've loaded it.

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:963]: I'm ambivalent about the arg order: this way is more aligned with what tokio::timeout looks like, but you have a fair point about async{} users.  Changed it.

Generally I don't put things in utils until they're really used more than one place, but this one feels like it inevitably will be, so sure.

ce82103a7

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1439]: I've played with making this one stricter by adding an explicit delete() and asserting in Drop that either delete or upsert was called.  The trouble is that it breaks the ergonomics of `?` in functions that hold a SlotGuard: they must avoid using `?` because if they hit an error then their slotguard gets illegally dropped without being written.

The kinds of errrors that we hit while holding a SlotGuard are generally very unexpected things, like failing to write a config file, where ultimately we'll probably make those infallible via fatal local io errors, and the current behavior in this PR of dropping the whole slot is a good one -- e.g. we were trying to attach a tenant and something weird went wrong, so we drop the slot.

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1518]: 2f405efb63783bbdcee6858e1822d62fad676fda

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1512]: This reminds me that we currently had no metrics at all for mgr.rs.  Added a couple to go with the unexpected errors on 2f405efb63783bbdcee6858e1822d62fad676fda

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1640]: Tucking the Ok path into the match makes sense though: 1b119d357

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs:1434]: I'm worried about dropping a tenant's slot before we're sure that all activity for that tenant has stopped.

This is kind of related to the Gate PR.

If we had async Drop, we could make the SlotGuard close the gate and wait for existing activity to cede.

Without it, IMO the SlotGuard should assert that the gate is closed and activity has ceded.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs:1292]: https://neondb.slack.com/archives/C033RQ5SPDH/p1699021103513689?thread_ts=1699013959.869989&cid=C033RQ5SPDH

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs]: > The ShuttingDown state never contains InProgress states -- shutdown_all_tenants0 doesn't carry them across.

Right, had missed that.

I'd prefer if you could make a separate `match` branch for the `TenantsMap::ShuttingDown` case and explain that in a comment on that branch.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs]: I see you won't budge, fair enough.

**Review by shanyp — APPROVED** [review]

File: pageserver/src/deletion_queue/check.log
Comment (arpad-m) [inline pageserver/src/deletion_queue/check.log:1]: this file should not be here

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs]: Done.

File: pageserver/src/http/routes.rs
Comment (arpad-m) [inline pageserver/src/http/routes.rs:165]: `e @ (AlreadyExists{..} | Conflict{..}) =>`

File: pageserver/src/tenant/mgr.rs
Comment (arpad-m) [inline pageserver/src/tenant/mgr.rs:1577]: could you instead make `if self.upserted { return}`?

**Review by arpad-m — CHANGES_REQUESTED** [review]

File: pageserver/src/tenant/mgr.rs
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1577]: b09891e52

File: pageserver/src/http/routes.rs
Comment (jcsp) [reply pageserver/src/http/routes.rs:165]: a758ed6179

**Review by arpad-m — APPROVED** [review]

**Comment by jcsp** [comment]
Test failure looks like a test helper issue:
- https://github.com/neondatabase/neon/issues/5794

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:1435]: No one questioned this?

Comment (arpad-m) [reply pageserver/src/tenant/mgr.rs:1435]: thanks for pointing this out, cross linking #5802 which addresses this.

**Pull request merged**
